src/webgpu/api/validation/queue/writeBuffer.spec.ts

```diff
@@ -11,10 +11,115 @@ Tests writeBuffer validation.
 
 Note: destroyed buffer is tested in destroyed/.
 
-TODO: implement.
+TODO: implement usage flag validation.
 `;
 
 import { makeTestGroup } from '../../../../common/framework/test_group.js';
+import { TypedArrayBufferView, TypedArrayBufferViewConstructor } from '../../../gpu_test.js';
 import { ValidationTest } from '../validation_test.js';
 
 export const g = makeTestGroup(ValidationTest);
+
+g.test('ranges')
+  .desc(
+    `
+Tests that the data ranges given to GPUQueue.writeBuffer() are properly validated. Tests calling
+writeBuffer with both TypedArrays and ArrayBuffers and checks that the data offset and size is
+interpreted correctly for both.
+
+  - When passing a TypedArray the data offset and size is given in elements.
+  - When passing an ArrayBuffer the data offset and size is given in bytes.
+
+Also verifies that the specified data range:
+
+  - Describes a valid range of the source buffer.
+  - Fits fully within the destination buffer.
+  - Is 4-byte aligned.
+`
+  )
+  .fn(async t => {
+    const queue = t.device.defaultQueue;
+
+    function runTest(arrayType: TypedArrayBufferViewConstructor, testBuffer: boolean) {
+      const elementSize = arrayType.BYTES_PER_ELEMENT;
+      const buffer = t.device.createBuffer({
+        size: 16 * elementSize,
+        usage: GPUBufferUsage.COPY_DST,
+      });
+      const arraySm: TypedArrayBufferView | ArrayBuffer = testBuffer
+        ? new arrayType(8).buffer
+        : new arrayType(8);
+      const arrayMd: TypedArrayBufferView | ArrayBuffer = testBuffer
+        ? new arrayType(16).buffer
+        : new arrayType(16);
+      const arrayLg: TypedArrayBufferView | ArrayBuffer = testBuffer
+        ? new arrayType(32).buffer
+        : new arrayType(32);
+
+      if (testBuffer) {
+        const array15 = new Uint8Array(15).buffer;
+
+        // Writing the full buffer that isn't 4-byte aligned.
+        t.shouldThrow('OperationError', () => queue.writeBuffer(buffer, 0, array15));
+
+        // Writing from an offset that causes source to be 4-byte aligned.
+        queue.writeBuffer(buffer, 0, array15, 3);
```

Comment:
Since this is a positive test, being maximally unaligned, as this is now, is best. Of course we will also want operation tests for corner cases like this.